keps/sig-network/1880-services-ip-ranges/README.md

```diff
@@ -0,0 +1,1030 @@
+
+#### The kube-apiserver bootstrap process and the service-cidr flags
+
+Currently, the Service CIDRs are configured independently in each kube-apiserver using flags. During
+the bootstrap process, the apiserver uses the first IP of each range to create the special
+"kubernetes.default" service. It also starts a reconcile loop, that synchronize the state of the
+bitmap with the assigned IPs to the Services.
+
+With current implementation, that each kube-apiserver can boot with different ranges configured.
+There is no conflict resolution, each apiserver keep writing and deleting others apiservers allocator bitmap
+and Services.
+
+In order to be completely backwards compatible, the bootstrap process will remain the same, the
+difference is that instead of creating a bitmap based on the flags, it will create new
+ServiceIPRange objects from the flags (flags configuration removal is out of scope of this KEP) 
+with an annotation `services.networking.kubernetes.io/is-default` set to true.
+
+With the new implementation only the Service IP Range configuration will be consistent, but not the
+default IP Family, that will still depend on the flags configuration of each individual apiserver.
+However, users still can provide consistency by using a webhook or admission controller for
+defaulting the IPFamily or IPFamilyPolicy fields on Services.
+
+There can be multiple default ServiceIPRange objects, also for the same IP family, since
+apiservers with different configuration can work, the only condition is that the defined
+service IP ranges contain the `kubernetes.default` Service.
+
+The new bootstrap process pseudocode will be:
+
+```sh
+parse --service-cluster-ip-range
+if CIDR flags are set
+	if a `default` range exists {
+		read it
+		if default != my flag values {
+      # current behavior doesn't stop multiple apiservers to boot with different configurations
+      # despite each one delete the services of the other
+			inform user: log warning/generate event (periodically)
+      use the `default` range configuration
+		}
+	} else { // no default range
+		try to create it from my flag values
+		if error {
+			if error == already exists {
+				// maybe racing with other apiserver
+				read it
+				if default != my flag values {
+          # current behavior doesn't stop multiple apiservers to boot with different configurations
+          # despite each one delete the services of the other
+          inform user: log warning/generate event (periodically)
+          use the `default` range configuration
+				}
+			} else {
+				handle error
+			}
+		}
+	}
+} else { // flags not set
+	if `default` range exists {
+		read it
+	} else { // no default range
+		handle error
+	}
+}
+```
+
+#### The special "default" ServiceIPRanges
+
+There has to be always one default ServiceIPRange containing the `kubernetes.default` Service.
+
+If the cluster is dual-stack, there has to be also one default ServiceIPRange per IP family.
```

Comment:
yeah, but I was thinking more about this on how users will use this and changed my mind
https://github.com/kubernetes/enhancements/pull/3365#discussion_r895142302

It is not the same situation as in nodes, pods are distributed uniformly across nodes so it makes sense to have the same masks per node (though it can change https://github.com/kubernetes/kubernetes/issues/109814#issuecomment-1138898983)


For Services, you assign an IPv6 /64 and you don't need more ranges anymore, we are making clear we don't want semantics on the ranges, Service IPs are fungible, so the real problem is in IPv4

And we are going to allow overlapping that means we have to add logic to the allocators to merge ranges, from the UX and implementation perspective it seems easier to think about single ranges, that was Casey point since the beginning https://github.com/kubernetes/enhancements/pull/3365#discussion_r892787939

Reply:
I updated comment there.  TL;DR - let's not block on this - add a beta criterion to revisit this and see if we are still happy?